File: cpl/ch4/stack.c
#include "calc.h"

/* 
 * Reverse Polish Calculator
 *
 * Based on code in Kernighan and Ritchie, second edition 
 */

enum { MAXVAL = 100 };

static int sp = 0;		/* next free stack position */
double val[MAXVAL];		/* value stack */

/* push: push f onto value stack */
void push(double f)
{

	if (sp < MAXVAL)
		val[sp++] = f;
	else
		fprintf(stderr, "error: stack full, cannot push %f\n", f);
}

/* pop: pop and return top value from stack */
double pop(void)
{

	if (sp > 0)
		return val[--sp];
	else 
		fprintf(stderr, "error: stack empty\n");

	return 0.0;		/* error */
}

/* peek: print top element of stack */
void peek(void)
{
	
	if (sp > 0)
		printf("%.8f\n", val[sp-1]);
	else 
		fprintf(stderr, "error: stack empty\n");
}

/* dup_top: duplicate top entry of stack */
void dup_top(void)
{
	
	if (sp > 0)
		push(val[sp-1]);
	else 
		fprintf(stderr, "error: stack empty\n");
}

/* swap: swap top two elements of stack */
void swap(void)
{
	double first, second;

	first = pop();
	second = pop();
	push(first);
	push(second);
}

/* clear: clear stack */
void clear(void)
{
	sp = 0;
}

File: gopl/ch07/xmltree/stack.go
// stack:  based on https://gist.github.com/bemasher/1777766
package xmltree

import "fmt"

type Stack struct {
	stack []Element
}

// Return the stack's length
func (sp *Stack) Len() int {
	return len(sp.stack)
}

// Push a new element onto the stack
func (sp *Stack) Push(n Element) {
	sp.stack = append(sp.stack, n)
}

// Remove the top element from the stack and return it's value
// If the stack is empty, return nil
func (sp *Stack) Pop() (Element, error) {
	var n Element

	if sp.Len() <= 0 {
		return n, fmt.Errorf("Stack empty")
	}

	lastIndex := sp.Len() - 1
	n = sp.stack[lastIndex]
	sp.stack = sp.stack[:lastIndex]
	return n, nil
}
